ddtrace/llmobs/_integrations/bedrock.py

```diff
@@ -144,6 +152,125 @@ def _extract_output_message_for_converse(response: Dict[str, Any]):
             return default_content
         return get_messages_from_converse_content(role, content)
 
+    @staticmethod
+    def _extract_output_message_for_converse_stream(
+        streamed_body: List[Dict[str, Any]]
+    ) -> Tuple[List[Dict[str, Any]], Dict[str, str], Dict[str, int]]:
+        """
+        Extract output messages from streamed converse responses.
+
+        Converse stream response comes in chunks. The chunks we care about are:
+        - a message start/stop event, or
+        - a content block start/stop event (for tool calls only currently)
+        - a content block delta event (for chunks of text in a message or tool call arg)
+        - usage metric information
+
+        For more info, see bedrock converse response stream response syntax:
+        https://docs.aws.amazon.com/bedrock/latest/APIReference/API_runtime_ConverseStream.html#API_runtime_ConverseStream_ResponseSyntax
+        """
+        usage_metrics: Dict[str, int] = {}
+        metadata: Dict[str, str] = {}
+        messages: List[Dict[str, Any]] = []
+
+        text_content_blocks: Dict[int, str] = {}
+        tool_content_blocks: Dict[int, Dict[str, Any]] = {}
+
+        current_message: Optional[Dict[str, Any]] = None
+
+        def process_message_end(current_message: Dict[str, Any]):
+            indices = sorted(current_message.get("context_block_indices", []))
+            message_output = {"role": current_message["role"]}
+
+            text_contents = [text_content_blocks[idx] for idx in indices if idx in text_content_blocks]
+            if text_contents:
+                message_output["content"] = "".join(text_contents)
+
+            tool_calls = []
+            for idx in indices:
+                if idx in tool_content_blocks:
+                    tool_block = tool_content_blocks[idx]
+                    tool_args = {}
+                    if "input" in tool_block:
+                        input_text = tool_block["input"]
+                        try:
+                            tool_args = json.loads(input_text)
+                        except (json.JSONDecodeError, ValueError):
+                            tool_args = {"input": input_text}
+
+                    tool_calls.append(
+                        {
+                            "name": tool_block.get("toolName", ""),
+                            "arguments": tool_args,
+                            "tool_id": tool_block.get("toolUseId", ""),
+                        }
+                    )
+
+            if tool_calls:
+                message_output["tool_calls"] = tool_calls
+
+            messages.append(message_output)
+
+        for chunk in streamed_body:
+            if "metadata" in chunk and "usage" in chunk["metadata"]:
+                usage = chunk["metadata"]["usage"]
+                if "inputTokens" in usage:
```

Comment:
This code is easy to turn into a loop over the sequence `("inputTokens", "outputTokens", "totalTokens")`, and doing so would increase its readability.